File: leetcode/cpp/combinationSum.cpp

// leetcode.com  39. Combination Sum

class Solution
{
  public:
    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {

        if (candidates.empty())
            return {};

        vector<vector<int>> res;
        vector<int> out;

        sort(candidates.begin(), candidates.end());

        helper(candidates, target, 0, out, res);

        return res;
    }

    void helper(vector<int> &candidates, int target, int level,
                vector<int> &out, vector<vector<int>> &res)
    {
        if (target < 0)
            return;

        if (target == 0)
        {
            res.push_back(out);
            return;
        }

        for (int i = level; i < candidates.size(); i++)
        {
            out.push_back(candidates[i]);
            helper(candidates, target - candidates[i], i, out, res);
            out.pop_back();
        }
    }
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/p000/CombinationSum.java
package com.vkeonline.leetcode.p000;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Leetcode [M]: 39. Combination Sum
 * @author csgear
 */
public class CombinationSum {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        candidates = Arrays.stream(candidates).sorted().distinct().toArray() ;
        List<List<Integer>> results = new ArrayList<>();
        dfs(candidates, 0, target, new ArrayList<>(), results);
        return results;
    }

    private void dfs(int[] candidates, int index, int target, List<Integer> combination, List<List<Integer>> results) {
        if (index == candidates.length) {
            if (target == 0) {
                results.add(new ArrayList<>(combination));
            }
            return;
        }

        if (target < 0) {
            return;
        }

        dfs(candidates, index + 1, target, combination, results);

        combination.add(candidates[index]);
        dfs(candidates, index, target - candidates[index], combination, results);
        combination.remove(combination.size() - 1);
    }
}
